### ISSUE TITLE
Unable to add cuda feature using cargo add

### ISSUE THREAD

**Issue by jaabberwocky** [huggingface/candle#1497]
I ran the command as per the tutorial:

`cargo add --git https://github.com/huggingface/candle.git candle-core --features "cuda"`

and got this error:

```
error: unrecognized feature for crate candle-core: cuda
no features available for crate candle-core
```

For diagnostic purposes:

1. `nvcc --version`

returns:

```
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Fri_Nov__3_17:16:49_PDT_2023
Cuda compilation tools, release 12.3, V12.3.103
Build cuda_12.3.r12.3/compiler.33492891_0
```

2. ` nvidia-smi --query-gpu=compute_cap --format=csv`

returns:

```
compute_cap
8.6
```

3. `nvidia-smi`

returns:

```
Fri Dec 29 16:02:02 2023
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.36                 Driver Version: 546.33       CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3080 Ti     On  | 00000000:2B:00.0  On |                  N/A |
|  0%   34C    P8              28W / 350W |   2135MiB / 12288MiB |     22%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A        31      G   /Xwayland                                 N/A      |
+---------------------------------------------------------------------------------------+
```

**Comment by grzuy** [comment]
I think you hit an apparently known recent cargo bug: https://github.com/rust-lang/cargo/issues/13121.

**Comment by grzuy** [comment]
Using `cargo add --features candle-core/cuda --git https://github.com/huggingface/candle.git candle-core` avoided the cargo issue for some reason...

**Comment by jaabberwocky** [comment]
Thank you @grzuy 👍  I can confirm that your solution works for me! Closing this as resolved.

**Comment by lamnguyenx** [comment]
> Using `cargo add --features candle-core/cuda --git https://github.com/huggingface/candle.git candle-core` avoided the cargo issue for me for some reason...

This method doesn't work for me.
I'm using :

```
Ubuntu 20.04.4 LTS
arch x86_64
rustc 1.75.0 (82e1608df 2023-12-21)
cargo 1.75.0 (1d8b05cdd 2023-11-20)
```

**Comment by corysabol** [comment]
> Using `cargo add --features candle-core/cuda --git https://github.com/huggingface/candle.git candle-core` avoided the cargo issue for me for some reason...

I also ran into this issue just now and this work around did not work for me.

Some diagnostics:

System: WSL2 Ubuntu

```
/usr/local/cuda/bin/nvcc --version
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2023 NVIDIA Corporation
Built on Wed_Nov_22_10:17:15_PST_2023
Cuda compilation tools, release 12.3, V12.3.107
Build cuda_12.3.r12.3/compiler.33567101_0
```
```
nvidia-smi --query-gpu=compute_cap --format=csv
compute_cap
8.9
```

```
nvidia-smi
Tue Jan 23 11:37:18 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 545.23.07              Driver Version: 546.12       CUDA Version: 12.3     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 4070 ...    On  | 00000000:01:00.0 Off |                  N/A |
| N/A   41C    P8               4W / 125W |    782MiB /  8188MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|  No running processes found                                                           |
+---------------------------------------------------------------------------------------+
```

It seems like cargo bug pointed at by @grzuy has been fixed, and should be in cargo 1.76.0. However, after a fresh stable install of the rust toolchain using rustup I'm still on cargo version 1.75.0

```
cargo --version
cargo 1.75.0 (1d8b05cdd 2023-11-20)
```

I quickly switched my toolchain to nightly using `rustup toolchain default nightly` which gave me cargo version `1.77.0-nightly`.
After this I was able to install candle with the cuda feature using the command in the installation instructions: 

```
cargo add --git https://github.com/huggingface/candle.git candle-core --features "cuda"
    Updating git repository `https://github.com/huggingface/candle.git`
      Adding candle-core (git) to dependencies.
             Features:
             + cuda
             + cudarc
             - accelerate
             - cudnn
             - metal
             - mkl
    Updating git repository `https://github.com/huggingface/candle.git`
    Updating crates.io index
```

So, this tells me that the fix for the issue may not have made it into the stable toolchain just yet. Hopefully this helps anyone else who stumbles upon this issue, even if it is already close.

**Comment by lambdald** [comment]
I also encountered the same problem. First, I used the command in the tutorial
`
cargo add --git https://github.com/huggingface/candle.git candle-core --features "cuda"
`
and encountered an error. Then I used the above command
`
cargo add --features candle-core/cuda --git https://github.com/huggingface/candle.git candle-core
`
and encountered an error as well. Finally, I removed the double quotation marks of cuda from command and the command ran successfully. The final command is
`
cargo add --git https://github.com/huggingface/candle.git candle-core --features cuda
`
